Extend the signaling setting API to allow somewhat clustering

nickvergessen opened the pull request:
- [x] Improve "random empty default picking"
- [x] Add token when requesting signaling information
- [x] Make sure the UI uses the correct signaling server per conversation

Fix #2591
Fix #3463

nickvergessen commented:
Should be functional now.

We still pick a random one for the first participant, but that is hard to make right without a real usecase in hand. But adjusting that should be rather easy in the end, so testing it for now should be fine.

Most effects should be noticable without the clustering.
But real clustering is skipped unless there is more then one signaling server defined and the config is set. Can be hacked in Config.php by adding `return self::SIGNALING_CLUSTER_CONVERSATION;` in `getSignalingMode()`.

By replacing: `currentSignalingServer = settings.url` with `currentSignalingServer = settings.url + 'different'` iin `src/utils/webrtc/index.js` you can also fake a different signalign server being used all the time, making you close and open the connection on every conversation change.

nickvergessen commented:
Would be nice if you can have a quick look and spill your opinion @danxuliu

nickvergessen commented:
Added a commit to check the config.php for a hardcoded list to allow overwriting the random for some known big conversations which can then be assigned specifically to always end up on the set machines

danxuliu commented on lib/TInitialState.php:
```diff
@@ -0,0 +1,125 @@
+		);
+
+		$signalingMode = $this->talkConfig->getSignalingMode();
+		if ($signalingMode !== 'internal' && !$this->memcacheFactory->isAvailable()) {
```
Is a distributed cache always needed for external signaling, or only for clustered external signaling?

In any case, I think this is too agressive. If an admin sets two external signaling servers and then happily navigates away from the settings to the Files app the admin will be greeted with that message. But then it will not be possible to open again the settings (at least, without using the direct URL) as trying to open the root Nextcloud URL will redirect to the Files app and the message will be shown again.

On the other hand, if setting several external signaling servers were prevented in the settings if a distributed cache is not set that would work when setting the servers, but if the distributed cache is then removed... That would be a problem. I am not sure how to handle that. So... I guess that the `HintException` is OKish :shrug: Yes, I am thinking out loud :-P

nickvergessen replied:
> Is a distributed cache always needed for external signaling, or only for clustered external signaling?

So it's not needed for the normal external signaling, but I wouldn't run a Nextcloud without it anyway if I need the HPB. But I added an appconfig now `signaling_dev` to not overwrite the signaling mode and not require the memcache

danxuliu commented on lib/Signaling/Manager.php:
```diff
@@ -0,0 +1,107 @@
+			return $servers[$serverId];
+		}
+
+		$serverIdToAssign = random_int(0, count($servers) - 1);
```
Missing try/catch.

nickvergessen replied:
Well yes, but actually if random_int fails with the HPB you have a sever problem anyway

danxuliu commented on lib/TInitialState.php:
```diff
@@ -50,6 +54,15 @@ protected function publishInitialStateShared(): void {
 		);
 
+		$signalingMode = $this->talkConfig->getSignalingMode();
+		if ($signalingMode !== 'internal'
```
Use `Config::SIGNALING_INTERNAL` instead.

nickvergessen commented:
> * Guests do not work (missing session ID from Make the signaling based on conversations)

Fixed

> * Forced reconnections do not work, at least with guests (it should not reload the page)

And fixed

> * Calls do not work in clustering mode (as they are initialized just once with the signaling server being used at that time and not changed later even if a different signaling connection/object is used)

I'm not sure what that is supposed to mean, care to explain?

nickvergessen commented:
Okay figured it, I had to actually join the calls.
Also fixed now.

And I will rebase once more due to #3495

danxuliu commented:
`webRtc.disconnect()` is not async, so it was not needed to use `await`. Besides that, `SimpleWebRTC.connection` is the signaling object, so calling `webRtc.disconnect()` and then `signaling.disconnect()` was redundant. However, although it should not be needed due to how the variables at the top of _webrtc.js_ are used, I modified `SimpleWebRTC.disconnect` to emit and event and reset those variables instead when disconnecting the SimpleWebRTC object.

nickvergessen commented:
> Independently of all that, when a participant leaves a call the participant is not removed from the call view of the rest of participants, but that is not related to this pull request.

That is only the last participant btw.

nickvergessen commented:
Ran unit tests locally, they pass and random integration tests also still worked, so merging

morph027 commented:
Jumping in here...Running latest version of NC and Talk (19 + 9) and Redis but i can't find a possibilty to add more than one signaling server in talk settings. Do i miss something?

nickvergessen commented:
The HPB can handle a few hundred users by default without any bigger problems. If you need the clustering you should get in contact with our sales team for a Nextcloud Talk Enterprise Subscription:

https://nextcloud.com/talk#pricing

morph027 commented:
Ah, well, not really necessary, just stumbled upon this in the release notes and wanted to try to satisfy my curiosity and improve knowledge about everything ;)
